rust/src/storage/s3/store.rs

```diff
@@ -0,0 +1,734 @@
+    ListObjectsV2Output,
+    ListObjectsV2Request,
+    ObjectIdentifier,
+    PutObjectError,
+    PutObjectOutput,
+    PutObjectRequest,
+    S3Client,
+    StreamingBody,
+    S3,
+};
+use std::{
+    collections::HashMap,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use thiserror::Error;
+use tokio::{io::AsyncReadExt, stream::Stream};
+
+#[derive(Debug, Error)]
+pub enum StorageError {
+    #[error("S3Error")]
+    S3Error,
+}
+
+#[derive(Clone)]
+struct Buckets(HashMap<&'static str, &'static str>);
+
+impl Buckets {
+    fn new() -> Self {
+        let mut names = HashMap::new();
+        names.insert("masks", "masks");
+        names.insert("masked-models", "masked-models");
+        names.insert("global-model", "global-model");
+        names.insert("global-masked-model", "global-masked-model");
+        Self(names)
+    }
+
+    fn masks(&self) -> &'static str {
+        self.0.get("masks").unwrap()
+    }
+
+    fn masked_models(&self) -> &'static str {
+        self.0.get("masked-models").unwrap()
+    }
+
+    fn global_model(&self) -> &'static str {
+        self.0.get("global-model").unwrap()
+    }
+
+    fn global_masked_model(&self) -> &'static str {
+        self.0.get("global-masked-model").unwrap()
+    }
+
+    fn names(&self) -> impl Iterator<Item = &&str> {
+        self.0.values()
+    }
+}
+
+#[derive(Clone)]
+pub struct S3Store {
+    s3_client: S3Client,
+    buckets: Buckets,
+}
+
+// API
+impl S3Store {
+    /// Create a new S3 store. The store creates and maintains 4 different buckets.
+    /// See [`Buckets`] for more information.
+    ///
+    /// The [`S3Store`] can also be used together with Minio via a custom region:
+    /// ```text
+    /// let region = Region::Custom {
+    ///    name: String::from("eu-east-3"),
+    ///    endpoint: String::from("http://127.0.0.1:9000"), // URL of minio
+    /// };
+    ///
+    /// let store = S3Store::new(region);
+    /// ```
+    pub fn new<P>(region: Region, credentials_provider: P) -> Self
+    where
+        P: ProvideAwsCredentials + Send + Sync + 'static,
+    {
+        let dispatcher = HttpClient::new().expect("failed to create request dispatcher");
+        Self {
+            s3_client: S3Client::new_with(dispatcher, credentials_provider, region),
+            buckets: Buckets::new(),
+        }
+    }
+
+    /// Upload a [`Mask`].
+    pub async fn upload_mask(&self, key: &str, mask: &Mask) -> Result<(), StorageError> {
+        match self
+            .upload(self.buckets.masks(), key, mask.serialize())
+            .await
+        {
+            Ok(_) => Ok(()),
+            Err(_) => Err(StorageError::S3Error),
+        }
+    }
+
+    /// Upload a [`MaskedModel`].
+    pub async fn upload_masked_model(
+        &self,
+        key: &str,
+        masked_model: &MaskedModel,
+    ) -> Result<(), StorageError> {
+        // we don't need the keys of the masked models. we can retrieve all keys via `list objects`.
```

Comment:
> we don't need the keys of the masked models

Do you mean that we never need to retrieve a specific mask? That's true, but I think there's still some value to be able to identify a mask. Perhaps we could use the update participant public key digest for that. But anyway it doesn't change anything for this method.